keras/utils/dataset_utils.py

```diff
@@ -14,14 +14,137 @@
 # ==============================================================================
 """Keras image dataset loading utilities."""
 
+
 import tensorflow.compat.v2 as tf
 # pylint: disable=g-classes-have-attributes
 
 import multiprocessing
 import os
+import time
 
 import numpy as np
+from tensorflow.python.util.tf_export import keras_export
+
+@keras_export('keras.utils.split_dataset')
+def split_dataset(dataset, 
+                  left_size=None, 
+                  right_size=None, 
+                  shuffle=False, 
+                  seed=None):
+  """Split a dataset into a left half and a right half (e.g. training / validation).
+  
+  Args:
+      dataset: A `tf.data.Dataset` object or 
+               a list/tuple of arrays with the same length.   
+      left_size: If float, it should be in range `[0, 1]` range 
+                 and signifies the fraction of the data to pack in 
+                 the left dataset. If integer, it signifies the number 
+                 of samples to pack in the left dataset. 
+                 If `None`, it defaults to the complement to `right_size`.       
+      right_size: If float, it should be in range `[0, 1]` range 
+                  and signifies the fraction of the data to pack 
+                  in the right dataset. If integer, it signifies 
+                  the number of samples to pack in the right dataset.
+                  If `None`, it defaults to the complement to `left_size`.     
+      shuffle: Boolean, whether to shuffle the data before splitting it.
+      seed: A random seed for shuffling.
+
+  Returns:
+      A tuple of two `tf.data.Dataset` objects: the left and right splits.
+  """
+  
+  if dataset and not isinstance(dataset,(tf.data.Dataset,list)):
+    raise TypeError('`dataset` must be either a tf.data.Dataset object'
+                     'or a list/tuple of arrays',f'Got {type(dataset)}')
+    
+  dataset_as_list = []
+  data_size_warning_flag = False
+  start_time = time.time()
+  
+  for datum in list(dataset):
```

Comment:
This `for` loop will not work in case where we have a list of arrays as `dataset`.

There should be a different loop for the array / list of arrays case and for the `Dataset` case.

Reply:
Note: adding unit test cases will help you get this right. Test for:

- `dataset=np.ones(shape=(200, 32))`
- `dataset=[np.ones(shape=(200, 32)), np.zeros(shape=(200, 32))]`
- `dataset=[np.ones(shape=(200, 32)), np.zeros(shape=(100, 32))]` -> should raise an error
- `dataset=(np.ones(shape=(200, 32)), np.zeros(shape=(200, 32)))`
- Batched tf.data.Dataset that yields batches of vectors
- Batched tf.data.Dataset that yields batches of tuples of vectors
- Batched tf.data.Dataset that yields batches of dicts of vectors
- Unbatched tf.data.Dataset that yields vectors
- Unbatched tf.data.Dataset that yields tuples of vectors
- Unbatched tf.data.Dataset that yields dicts of vectors